active_discussion/embedded/static-initialization.md

```diff
@@ -0,0 +1,137 @@
+# Runtime initialization of static variables
+
+Sometimes one needs to initialize a `static mut` variable, that will be
+accessed by an interrupt handler, at *runtime*. For this scenario we have a
+pattern where the variable starts uninitialized in const context (e.g. `static
+X: _ = MaybeUninit::uninit()`) and then it's initialized *before* the first
+invocation of the interrupt handler. We use a compiler fence or the "memory"
+clobber to enforce strict ordering between the initialization of the static
+variable and the first invocation of the interrupt handler, but is that enough?
+
+Note that:
+
+- The programs in this document target the [Basic Embedded Abstract Machine
+  (BEAM)][beam]. Please become familiar with the linked specification before you
+  read the rest of this document.
+
+[beam]: https://github.com/rust-lang/unsafe-code-guidelines/pull/111
+
+- In these programs we assume that [rust-lang/rfcs#2585][rfc2585] has been
+  accepted and implemented.
+
+[rfc2585]: https://github.com/rust-lang/rfcs/pull/2585
+
+## Memory clobber synchronization
+
+Consider this program
+
+``` rust
+#![no_std]
+
+use core::mem::MaybeUninit;
+
+static mut X: MaybeUninit<bool> = MaybeUninit::uninit();
+
+#[no_mangle]
+unsafe fn main() -> ! {
+    X.write(false);
+
+    unsafe {
+        asm!("ENABLE_INTERRUPTS" : : : "memory" : "volatile");
+        //                             ^^^^^^^^
+    }
+
+    // `INTERRUPT0` can preempt `main` from this point on and at any time
+
+    loop {
+        // .. any safe code ..
+    }
+}
+
+#[no_mangle]
+unsafe fn INTERRUPT0() {
+    let x: &mut bool = unsafe { &mut *X.as_mut_ptr() };
+
+    // .. any safe code ..
+}
+```
+
+Note that "any safe code" can *not* call `main` or `INTERRUPT0` (because they
+are `unsafe` functions), use `asm!` or access registers.
+
+**Claim**: the memory clobber is sufficient to prevent misoptimizations.
+
+"Why is the memory clobber required?" If the compiler reorders `X.write` to
+after the `asm!` block `INTERRUPT0` could observe `X` in an uninitialized state.
+
```

Comment:
You might be slightly better off rewriting this example to use a `core::mem::black_box(X.as_mut_ptr())` that has some guaranteed semantics, but AFAICT that would be an intrinsic that's a `nop` in the virtual machine, but somehow is guaranteed to disable some optimizations, which is a discussion that did not go very well last time.

Otherwise, it feels that modeling this would require a Rust memory model that understands inline assembly.

Reply:
I have not been actively following that RFC but I recall reading that using `black_box` for memory safety was discouraged ("it should only be used for benchmarks", is what I recall). Is that no longer the case?

Reply:
There is no RFC for a `core::mem::black_box` operation with guaranteed semantics. Somebody would need to write one. There is an RFC for a `core::hint::black_box` intrinsics open, but that is not what I was suggesting to use here.

Since inline assembly is not a part of the Rust language (as seen in the other threads, it's not even clear what `volatile`/`memory` as clobbers do / should do in Rust), I don't think discussing anything related to inline assembly is a good use of this working group's time at this point.